File: week 1/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
  /* prompt input of CC number and store in long long */
  printf("Number: ");
  long long ccNumber = get_long_long();
  
  /* validate  that number is greater than 0 or prompt for retry */
  while (ccNumber <= 0) {
    printf("Retry: ");
    ccNumber = get_long_long();
  }
  
  int count = 0;
  int firstInt = 0;
  
  /* initialize temp number equal to ccNumber to get number length */
  long long testDigits = ccNumber;
  while (testDigits > 0) {
    testDigits /= 10;
    count++;
  }
  
  /* place each digit into array */
  int numberArray[count];
  for (int i = 0; i < count; i++) {
    numberArray[i] = ccNumber % 10;
    ccNumber /= 10;
  }
  
  /* store first integer of number for use in validating visa vs. mastercard numbers */
  firstInt = numberArray[count-1];
  
  /* multiply every other number starting with 2nd to last number */
  for (int i = 1; i < count; i+=2) {
    numberArray[i] *= 2;
  }
    
  int sum = 0;
  /* add individual digits of products */  
  for (int i = 1; i < count; i+=2) {
    if (numberArray[i] > 9) {
      sum += 1 + (numberArray[i] - 10);
    } else {
      sum += numberArray[i];
    }
  }
  
  /* add other digits to sum */
  for (int i = 0; i < count; i+=2) {
    sum += numberArray[i];
  }
  
  /* validate number and determine company */
  if (!(sum % 10) == 0) {
    printf("INVALID\n");
  } else if (count == 15) {
    printf("AMEX\n");
  } else if (count == 16 && firstInt == 4) {
    printf("VISA\n");
  } else if (count == 13 || count == 16) {
    printf("MASTERCARD\n");
  }
}

File: Week 8/credit.py
import cs50

def main():
    
    # Prompt for credit card number input using cs50's get_int function to guarantee input is type int
    print("Number: ", end="")
    ccNumber = cs50.get_int()
    
    # Prompt for "Retry: " while input number is less than or equal to 0
    while ccNumber <= 0:
        print("Retry: ", end="")
        ccNumber = cs50.get_int()
    
    # Convert ccNumber to str to use len method to get number of integers    
    count = len(str(ccNumber))
    
    # Create empty list and add individual digits to list for arimethic manipulation as prescribed by algorithm    
    numberArray = []
    for i in range(count):
        numberArray.append(ccNumber % 10)
        ccNumber //= 10
    
    # Remember first integer for use in determining type of card between Visa and MasterCard which can both have 16 digits
    firstInt = numberArray[count-1]
    
    # Multiply numbers in odd indices by two, starting with 2nd to last number (numbers were added to array in reverse order using append)
    for i in range(count):
        if i % 2 != 0:
            numberArray[i] *= 2

    # Initialize variable sum for global use and add values in odd indices, making sure to subtract 10 if num is greater than 10
    sum = 0
    for i in range(count):
        if i % 2 != 0:
            if numberArray[i] > 9:
                sum += 1 + (numberArray[i] - 10)
            else:
                sum += numberArray[i]
    # Add values in even indices to sum
    for i in range(count):
        if i % 2 == 0:
            sum += numberArray[i]
    
    # Test number for invalidity. If valid, use if/else statements to check for type of card and print type
    if sum % 10 != 0:
        print("INVALID")
    elif count == 15:
        print("AMEX")
    elif count == 16 and firstInt == 4:
        print("VISA")
    elif count == 13 or count == 16:
        print("MASTERCARD")
        
        
if __name__ == "__main__":
    main()
    
    
    